### DISCUSSION TITLE
How to best port flask apps to v3 ?

### DISCUSSION THREAD

**Discussion by paketb0te** [spec-first/connexion#1802] [Q&A] [answered]
Hey all,

with the recent release of connexion 3.0, [apps now need to be run with an ASGI server](https://connexion.readthedocs.io/en/latest/v3.html#asgi-server).
I appreciate this, but have a question regarding the use of `FlaskApp`:

we have a bunch of flask apps where we do not use any async functions (which works ok in our current setup with WSGI). Our current setup ist to run the app in Apache, using uWSGI (where parallel request handling is achieved by multiple processes / threads).

How does that translate to running the app in an ASGI server?

Is there some mechanism that takes care of handling multiple parallel requests (e.g. dispatching requests to separate threads, like [`asyncio.to_thread`](https://docs.python.org/3/library/asyncio-task.html#asyncio.to_thread)), or do I need to take care of that myself by passing such parameters to the webserver running the app?

Assuming that we cannot (in the short term) port all apps to be actually async, what is the correct way to run blocking functions in connexion 3 / ASGI ?

I'm thankful for any pointers, I could not find anything related in the [Running your application](https://connexion.readthedocs.io/en/latest/quickstart.html#running-your-application) section of the docs.

**Comment by RobbeSneyders** [comment, marked as answer]
Hi @paketb0te,

For processes, you can set the number of workers in the ASGI server.

Threads depends on the Connexion Application you use:
- `FlaskApp`: Since `Flask` is a `WSGI` app instead of `ASGI`, we execute any requests passed to the `FlaskApp` in a thread pool (of 10 workers) automatically. If you define an `async` routes in the `FlaskApp`, you're just using [`Flask` async routes](https://flask.palletsprojects.com/en/2.3.x/async-await/) within a thread.
-  `AsyncApp`: The `AsyncApp` implements ASGI, so here threads are completely replaced by `asyncio`.

**Reply by RobbeSneyders** [reply]
A short term fix could be to [wrap Connexion with an ASGI-to-WSGI middleware](https://connexion.readthedocs.io/en/latest/v3.html#asgi-server), although I'd advice against that.

**Reply by paketb0te** [reply]
Thanks for clarifying, @RobbeSneyders !

We currently use `FlaskApp`, so it's good to know that Connexion uses a thread pool automatically. Is my conclusion correct, that this means a single ASGI server process can handle 10 requests in parallel?

And regarding a possible transition to `AsyncApp`: Is my understanding correct, that it is _not_ helpful to just use the same blocking code inside an `AsyncApp`?
